test_conformance/subgroups/workgroup_common_templates.h

```diff
@@ -0,0 +1,778 @@
+                            return TEST_FAIL;
+                        }
+                    }
+                }
+            }
+            x += nw;
+            y += nw;
+            m += 4 * nw;
+        }
+        log_info("  sub_group_%s(%s)... passed\n", operation_names(operation),
+                 TypeManager<Ty>::name());
+        return TEST_PASS;
+    }
+};
+
+static float to_float(subgroups::cl_half x) { return cl_half_to_float(x.data); }
+
+static subgroups::cl_half to_half(float x)
+{
+    subgroups::cl_half value;
+    value.data = cl_half_from_double(x, CL_HALF_RTE);
+    return value;
+}
+
+// for integer types
+template <typename Ty> inline Ty calculate(Ty a, Ty b, ArithmeticOp operation)
+{
+    switch (operation)
+    {
+        case ArithmeticOp::add_: return a + b;
+        case ArithmeticOp::max_: return a > b ? a : b;
+        case ArithmeticOp::min_: return a < b ? a : b;
+        case ArithmeticOp::mul_: return a * b;
+        case ArithmeticOp::and_: return a & b;
+        case ArithmeticOp::or_: return a | b;
+        case ArithmeticOp::xor_: return a ^ b;
+        case ArithmeticOp::logical_and: return a && b;
+        case ArithmeticOp::logical_or: return a || b;
+        case ArithmeticOp::logical_xor: return !a ^ !b;
+        default: log_error("Unknown operation request"); break;
+    }
+    return 0;
+}
+// Specialize for floating points.
+template <>
+inline cl_double calculate(cl_double a, cl_double b, ArithmeticOp operation)
+{
+    switch (operation)
+    {
+        case ArithmeticOp::add_: {
+            return a + b;
+        }
+        case ArithmeticOp::max_: {
+            return a > b ? a : b;
+        }
+        case ArithmeticOp::min_: {
+            return a < b ? a : b;
+        }
+        case ArithmeticOp::mul_: {
+            return a * b;
+        }
+        default: log_error("Unknown operation request"); break;
+    }
+    return 0;
+}
+
+template <>
+inline cl_float calculate(cl_float a, cl_float b, ArithmeticOp operation)
+{
+    switch (operation)
+    {
+        case ArithmeticOp::add_: {
+            return a + b;
+        }
+        case ArithmeticOp::max_: {
+            return a > b ? a : b;
+        }
+        case ArithmeticOp::min_: {
+            return a < b ? a : b;
+        }
+        case ArithmeticOp::mul_: {
+            return a * b;
+        }
+        default: log_error("Unknown operation request"); break;
+    }
+    return 0;
+}
+
+template <>
+inline subgroups::cl_half calculate(subgroups::cl_half a, subgroups::cl_half b,
+                                    ArithmeticOp operation)
+{
+    switch (operation)
+    {
+        case ArithmeticOp::add_: return to_half(to_float(a) + to_float(b));
+        case ArithmeticOp::max_:
+            return to_float(a) > to_float(b) || is_half_nan(b.data) ? a : b;
+        case ArithmeticOp::min_:
+            return to_float(a) < to_float(b) || is_half_nan(b.data) ? a : b;
```

Comment:
I'm not sure we can have NaN as inputs. See my comment below. I'm not sure whether this function should be impacted by this though.